Add external glusterfs backend to hosted and glusterfs playbook

mjudeikis opened the pull request:
This allows configuring openshift registry to use external gluster as storage backend.

TODO:
- [x]  Add checks if variables are set together and give error. as [1] variables is not compatible with [2] groups
- [x] Update readme with how to configure external gluster
- [x] Update openshift doc with external gluster configuration
- [x] test before merge
- [x] run tests and linting 


[1]
```
openshift_hosted_registry_storage_glusterfs_path=test_cluster
openshift_hosted_registry_storage_glusterfs_endpoints=ccr-gluster
openshift_hosted_registry_storage_glusterfs_ips=192.168.20.239,192.168.20.96,192.168.20.114
```
[2]
```
[glusterfs]
[glusterfs_registry]
```

FIY:  @jarrpa @IshentRas

mjudeikis commented:
inventory files for 2 connected env where I run tests:
 https://gist.github.com/mjudeikis/18c82fdd32bc788997ff2e4975503ca2

mjudeikis commented:
/assign @jarrpa

mjudeikis commented on roles/openshift_storage_glusterfs/tasks/heketi_deploy_part2.yml:
```diff
@@ -1,6 +1,6 @@
 - name: Create heketi DB volume
-  command: "{{ glusterfs_heketi_client }} setup-openshift-heketi-storage --image {{ glusterfs_heketi_image}}:{{ glusterfs_heketi_version }} --listfile /tmp/heketi-storage.json"
+  command: "{{ glusterfs_heketi_client }} setup-openshift-heketi-storage --listfile /tmp/heketi-storage.json"
   register: setup_storage
```
This is bug. New version of heketi does not need image anymore

mjudeikis replied:
It fails with "image flag is unknown". are you sure this is needed?

jarrpa replied:
I should probably clarify: yes this should get fixed, just in a separate PR. :)

mjudeikis commented:
/test tox
/test install

mjudeikis commented:
/retest

mjudeikis commented:
somehow my both test env succ and our install fails on last step to remove temp folder we created in registry.yml. checking...

jarrpa commented on roles/openshift_hosted/README.md:
```diff
@@ -55,6 +63,11 @@ Example Playbook
     openshift_hosted_router_registryurl: 'registry.access.redhat.com/openshift3/ose-haproxy-router:v3.0.2.0'
     openshift_hosted_router_selector: 'type=infra'
+    openshift_hosted_registry_storage_kind=glusterfs
+    openshift_hosted_registry_storage_glusterfs_path=external_registry_path
```
Change the value to something like `external_glusterfs_volume_name`

mjudeikis replied:
this one comes from oo_filters.py .lets keep it as it is if possible. Will check how I can reuse it smoothly.

jarrpa replied:
I didn't mean the variable name, I meant the example value you're assigning to it.

mjudeikis replied:
ou :) k

jarrpa commented on roles/openshift_hosted/README.md:
```diff
@@ -55,6 +63,11 @@ Example Playbook
     openshift_hosted_router_selector: 'type=infra'
+    openshift_hosted_registry_storage_kind=glusterfs
+    openshift_hosted_registry_storage_glusterfs_path=external_registry_path
+    openshift_hosted_registry_storage_glusterfs_endpoints=external-gluster-ep-name
```
You don't need to set this. The default of `glusterfs-registry-endpoints` should suffice, since our constraints make it such that we don't allow a conflict.

jarrpa commented on roles/openshift_hosted/defaults/main.yml:
```diff
@@ -71,7 +71,8 @@ openshift_hosted_registry_name: docker-registry
 openshift_hosted_registry_cert_expire_days: 730
-
+openshift_hosted_registry_external_gluster: False
+openshift_hosted_registry_storage_glusterfs_endpoints: 'glusterfs-registry-endpoints'
```
This is taken care of in the filter plugin. You don't need to specify defaults here.

jarrpa commented on roles/openshift_hosted/tasks/registry.yml:
```diff
@@ -43,6 +49,17 @@
     openshift_hosted_registry_namespace: "{{ openshift.hosted.registry.namespace | default('default') }}"
     openshift_hosted_registry_selector: "{{ openshift.hosted.registry.selector }}"
     openshift_hosted_registry_images: "{{ openshift.hosted.registry.registryurl | default('openshift3/ose-${component}:${version}')}}"
+    openshift_hosted_registry_storage_glusterfs_path: "{{ openshift_hosted_registry_storage_glusterfs_path | default('') }}"
```
Specify this in the `defaults/main.yml` file.

jarrpa commented on roles/openshift_hosted/tasks/registry.yml:
```diff
@@ -43,6 +49,17 @@
     openshift_hosted_registry_selector: "{{ openshift.hosted.registry.selector }}"
     openshift_hosted_registry_images: "{{ openshift.hosted.registry.registryurl | default('openshift3/ose-${component}:${version}')}}"
+    openshift_hosted_registry_storage_glusterfs_path: "{{ openshift_hosted_registry_storage_glusterfs_path | default('') }}"
+    openshift_hosted_registry_storage_glusterfs_ips: "{{ openshift_hosted_registry_storage_glusterfs_ips | default('') }}"
```
Specify this in the `defaults/main.yml` file.

jarrpa commented on roles/openshift_sanitize_inventory/tasks/unsupported.yml:
```diff
@@ -32,3 +32,15 @@
       openshift_cloudprovider_kind. You can disable this check with
       'dynamic_volumes_check=False'.
+
+- name: Ensure that gluster CNS/CRS/External Gluster are configured right
```
Rewrite this to something like: `Ensure the hosted registry's GlusterFS storage is configured correctly`

jarrpa commented on roles/openshift_sanitize_inventory/tasks/unsupported.yml:
```diff
@@ -32,3 +32,15 @@
+    msg: |-
+      Using a registry storage kind of 'glusterfs' with variable openshift_hosted_registry_storage_glusterfs_ips and host groups glusterfs/glusterfs_registry is not possible.
+      If glusterfs/glusterfs_registry groups are set in the invetory file, registry backend is configured automaticaly (CNS and CRS). On the other side openshift_hosted_registry_storage_glusterfs_ips implies
+      that you are configuring external preconfigured gluster for registry storage backend.
```
Rewrite this to something like:
`Configuring a value for openshift_hosted_registry_storage_glusterfs_ips and with a glusterfs_registry host group is not allowed. Specifying a glusterfs_registry host group indicates that a new GlusterFS cluster should be configures, whereas specifying openshift_hosted_registry_storage_glusterfs_ips indicates wanting to use a pre-configured GlusterFS cluster for the registry storage.`

jarrpa commented on roles/openshift_storage_glusterfs/tasks/glusterfs_registry.yml:
```diff
@@ -27,24 +27,13 @@
     glusterfs_heketi_ssh_sudo: "{{ openshift_storage_glusterfs_registry_heketi_ssh_sudo | bool }}"
     glusterfs_heketi_ssh_keyfile: "{{ openshift_storage_glusterfs_registry_heketi_ssh_keyfile }}"
-    glusterfs_nodes: "{{ groups.glusterfs_registry | default(groups.glusterfs) }}"
+    glusterfs_nodes: "{% if groups.glusterfs_registry is defined %}{% set nodes = groups.glusterfs_registry %}{% elif 'groups.glusterfs' is defined %}{% set nodes = groups.glusterfs %}{% else %}{% set nodes = '[]' %}{% endif %}{{ nodes }}"
```
You don't need to use the interim `nodes` variable, just output the desired values between the `{% if %}` clauses.

mjudeikis commented on roles/openshift_storage_glusterfs/tasks/main.yml:
```diff
@@ -11,7 +11,7 @@
 - include: glusterfs_registry.yml
   when:
-  - "groups.glusterfs_registry | default([]) | count > 0 or openshift.hosted.registry.storage.kind == 'glusterfs' or openshift.hosted.registry.storage.glusterfs.swap"
+  - "groups.glusterfs_registry | default([]) | count > 0  or openshift.hosted.registry.storage.kind == 'glusterfs' or openshift.hosted.registry.storage.glusterfs.swap"
```
wups :D

mjudeikis commented:
@jarrpa Check again please. Not everything as per your comments, but looks much better now (I think). Still doing some testing on my [laptop](https://twitter.com/Mangirdas/status/918498857251729409).

Removed EP,SVC creation from glusterfs role. But I would like to stick with oc apply instead of oc_obj, as `oc apply` allows role idempotent without deleting and recreating things.  It think `oc_obj` should be refactored to use `oc apply`, instead of `oc create`

jarrpa commented on roles/openshift_hosted/tasks/registry.yml:
```diff
@@ -49,17 +49,7 @@
-  - openshift_hosted_registry_storage_glusterfs_path != ''
-  - openshift_hosted_registry_storage_glusterfs_ips != ''
-  - not 'glusterfs' in groups | default([])
+    openshift_hosted_registry_storage_glusterfs_ips: "{% set comma = joiner(',') %}{% if groups.glusterfs_registry is defined %}{% for node in groups.glusterfs_registry %}{{ comma() }}{{  hostvars[node].glusterfs_ip | default(hostvars[node].openshift.common.ip)}}{% endfor %}{% elif groups.glusterfs is defined %}{% for node in groups.glusterfs %}{{ comma() }}{{ hostvars[node].glusterfs_ip | defaul ...
```
A few things here:
* The current order of precedence is `glusterfs_registry` -> `glusterfs` -> `glusterfs_ips`. Do you think it would make sense to move `glusterfs_ips` as the first choice?
* Is the 'joiner' method preferred to using `{% if not loop.first %}` or something? I don't have an opinion on that, I just haven't seen `joiner` before.
* Add a space after `default(hostvars[node].openshift.common.ip)` :)

mjudeikis replied:
@jarrpa i think order is fine now. As it populate variables later.

For joiner - will do some tests. Had some issues with standart join due format comming from ansible.

mjudeikis replied:
Tests fails as I missed one usecase. Will fix later

jarrpa replied:
I meant something like `loop.last` in jinja2: http://jinja.pocoo.org/docs/2.9/templates/#for

mjudeikis replied:
If you want I can change to loop.last but Joiner looks more nicer for me :D as it requires less if else conditions.

jarrpa replied:
Nah, that's fine! I'd just never seen it before. :) I'm fine leaving it as-is.

mjudeikis commented on roles/openshift_storage_glusterfs/tasks/heketi_deploy_part2.yml:
```diff
@@ -1,6 +1,6 @@
 - name: Create heketi DB volume
-  command: "{{ glusterfs_heketi_client }} setup-openshift-heketi-storage --image {{ glusterfs_heketi_image}}:{{ glusterfs_heketi_version }} --listfile /tmp/heketi-storage.json"
+  command: "{{ glusterfs_heketi_client }} setup-openshift-heketi-storage --listfile /tmp/heketi-storage.json"
   register: setup_storage
```
I think it is fixed now in heketi

mjudeikis commented:
@jarrpa I think Im done for this round. I have few other changes, but those will go to separate PR

jarrpa commented:
@mjudeikis Scenario 3 is a Pass. It's perfectly reasonable to set up the glusterfs_registry cluster and the registry volume in preparation for deploying a reigstry later. I'll give the latest code a review today.

mjudeikis commented:
/test upgrade

jarrpa commented on roles/openshift_storage_glusterfs/tasks/glusterfs_config.yml:
```diff
@@ -27,6 +27,6 @@
     glusterfs_heketi_ssh_sudo: "{{ openshift_storage_glusterfs_heketi_ssh_sudo | bool }}"
     glusterfs_heketi_ssh_keyfile: "{{ openshift_storage_glusterfs_heketi_ssh_keyfile }}"
-    glusterfs_nodes: "{{ groups.glusterfs }}"
+    glusterfs_nodes: "{{ groups.glusterfs | default(0) }}"
```
Default to `[]`.

jarrpa commented on roles/openshift_hosted/tasks/storage/glusterfs_external.yml:
```diff
@@ -0,0 +1,16 @@
+    src: "{{ openshift.common.examples_content_version }}/glusterfs-registry-endpoints.yml.j2"
+    dest: "{{ mktempHosted.stdout }}/glusterfs-registry-endpoints.yml"
+
+- name: Copy GlusterFS registry service
```
s/Copy/Generate/

mjudeikis commented:
> Could you just take the contents of glusterfs_external.yml and insert them into glusterfs.yml? It seems like they should always be run regardless of GlusterFS configuration

If we want to merge it to glusterfs.yml we need to include it twice. As we need to create EP and SVC before we even start registry pod, and all mounting, SELinux context settings (`registry_pods.results.results[0]['items'][0].spec.securityContext.fsGroup`) have to happen after we create DC. I would make it something like `gluster_pre.yml` and `gluster_post.yml` or include the same file but add some conditional variables to which part we want to execute. But in this case I prefer 2 files - Keep it simple, stupid - KISS :)

mjudeikis commented:
/test install

jarrpa commented:
@mjudeikis I don't entirely see what you mean. I only see three tasks between the two includes. Is the `oc_adm_registry` command more complex than I think?

If there is a need for that split I'm not seeing, yeah rename them. :) Maybe something simple like `glusterfs_part1` and `glusterfs_part2`.

mjudeikis commented:
@jarrpa it mainly because of this:
https://github.com/openshift/openshift-ansible/blob/master/roles/openshift_hosted/tasks/storage/glusterfs.yml#L24 

Will rename this stuff or check whats can be done later :)

jarrpa commented:
@mjudeikis ...okay, so why not just bring the "external" stuff at the beginning of THAT file, and include that file after the registry pods exist?

mjudeikis commented:
@jarrpa ping me on IRC :)

jarrpa commented on roles/openshift_hosted/README.md:
```diff
@@ -33,13 +33,20 @@ variables also control configuration behavior:
+| openshift_hosted_registry_storage_glusterfs_swap     | False         | Whether to swap an existing registry's storage volume for a GlusterFS volume |
+| openshift_hosted_registry_storage_glusterfs_swapcopy | True          | If swapping, also copy the current contents of the registry volume           |
+| openshift_hosted_registry_storage_glusterfs_ips      | []            | If we use external gluster for Registry backend we need to provide list of IP's for EP creation|
+| openshift_hosted_registry_storage_glusterfs_path     | []            | If we use external gluster we need to provide volume name from gluster cluster for mounting|
```
Do not add this, this is covered in the GlusterFS role.

mjudeikis replied:
If user do not deploy gluster and want just to use it, why he/she should go and read readme of glsuterFS role? :) My vote is to keep it here too.

jarrpa commented on roles/openshift_hosted/README.md:
```diff
@@ -33,13 +33,20 @@ variables also control configuration behavior:
-| openshift_hosted_registry_glusterfs_swapcopy | True          | If swapping, also copy the current contents of the registry volume           |
+| openshift_hosted_registry_storage_glusterfs_swap     | False         | Whether to swap an existing registry's storage volume for a GlusterFS volume |
+| openshift_hosted_registry_storage_glusterfs_swapcopy | True          | If swapping, also copy the current contents of the registry volume           |
+| openshift_hosted_registry_storage_glusterfs_ips      | []            | If we use external gluster for Registry backend we need to provide list of IP's for EP creation|
```
Change the description to: `A list of IP addresses for the GlusterFS cluster to use for hosted registry storage`

jarrpa commented on roles/openshift_hosted/README.md:
```diff
@@ -33,13 +33,20 @@ variables also control configuration behavior:
 ------------
 
 * openshift_hosted_facts
+* openshift_storage_glusterfs
```
This isn't true. It's not a dependency if you're not using GlusterFS storage. Remove it.

mjudeikis commented:
@jarrpa ,

>In reviewing your use of the glusterfs_ips variable, it would make much more sense for it to be a list. I would suggest altering the code to treat it as a list and just add an early clause like if type(glusterfs_ips) == string to convert the string into a list. That way users can still specify a simple list.

Lets keep it either list or string. Changed to list. Easier to maintain one state of variable

>The glusterfs_ips default should, hilariously, not be specifie din the defaults file. It should follow the pattern of the other defaults and be defined here. I'm not sure what the reason is for those being there, but I'd rather be consistent with preexisting patterns until they are corrected. :)

Done

>Related to that last point, with the changed default location the glusterfs_ips variable should change to 
use periods instead of underscores like the other glusterfs variables.

Moved there appropriate. 

```
{%- set gluster_ips = [] %}
{% if groups.glusterfs_registry is defined %}
    {% for node in groups.glusterfs_registry %}
        {%- set _ = gluster_ips.append(hostvars[node].glusterfs_ip | default(hostvars[node].openshift.common.ip)) %}
    {% endfor %}
    {{ gluster_ips }}
{% elif groups.glusterfs is defined %}
    {% for node in groups.glusterfs %}
       {%- set _ = gluster_ips.append(hostvars[node].glusterfs_ip | default(hostvars[node].openshift.common.ip)) %}
    {% endfor %}
    {{ gluster_ips }}
{% else %}
    {{ openshift.hosted.registry.storage.glusterfs.ips }}
{% endif %}
```
This can be done in like 5 different ways in a shorter way - but I rather keep it "human readable" syntax, than some filter when somebody else needs to do 2-3h of reverse engineering on logic. Its already overkill  :/ 

I hope we can call it done for this one :D :D

mjudeikis commented:
/test logging

jarrpa commented on roles/openshift_hosted/README.md:
```diff
@@ -33,13 +33,17 @@ variables also control configuration behavior:
 | Name                                         | Default value | Description                                                                  |
 |----------------------------------------------|---------------|------------------------------------------------------------------------------|
-| openshift_hosted_registry_glusterfs_swap     | False         | Whether to swap an existing registry's storage volume for a GlusterFS volume |
-| openshift_hosted_registry_glusterfs_swapcopy | True          | If swapping, also copy the current contents of the registry volume           |
```
I completely disagree with this removal. If we are documenting the ips and path values here, all `reigistry_strorage_glusterfs` variables should be documented here. Copy all values from `roles/openshift_storage_glusterfs/README.md` into here and change said file to reference this one with something like the following:

`Additionally, this role's behavior responds to several registry-specific variables in the [openshift_hosted role](RELATIVE_URL).`

jarrpa commented on roles/openshift_hosted/README.md:
```diff
@@ -33,13 +33,17 @@ variables also control configuration behavior:
 |----------------------------------------------|---------------|------------------------------------------------------------------------------|
-| openshift_hosted_registry_glusterfs_swap     | False         | Whether to swap an existing registry's storage volume for a GlusterFS volume |
-| openshift_hosted_registry_glusterfs_swapcopy | True          | If swapping, also copy the current contents of the registry volume           |
+| openshift_hosted_registry_storage_glusterfs_ips      | []            |A list of IP addresses for the GlusterFS cluster to use for hosted registry storage|
```
Put the brackets in `` ` `` backticks.

jarrpa commented on roles/openshift_hosted/README.md:
```diff
@@ -33,13 +33,17 @@ variables also control configuration behavior:
+| openshift_hosted_registry_storage_glusterfs_ips      | []            |A list of IP addresses for the GlusterFS cluster to use for hosted registry storage|
+| openshift_hosted_registry_storage_glusterfs_path     | []            |If we use external gluster we need to provide volume name from gluster cluster for mounting|
+
+**Note:** if you have host group `glusterfs_registry` set in your invetory, variable `openshift_hosted_registry_storage_glusterfs_ips`, CAN'T be used because this is 2 different use-cases - CNS/CRS vs External Native Gluster. 
```
Use the wording we already have elsewhere:

```
**Note:** Configuring a value for 
openshift_hosted_registry_storage_glusterfs_ips and with a glusterfs_registry 
host group is not allowed. Specifying a glusterfs_registry host group 
indicates that a new GlusterFS cluster should be configured, whereas 
specifying openshift_hosted_registry_storage_glusterfs_ips indicates wanting 
to use a pre-configured GlusterFS cluster for the registry storage.
```
I don't want product names in this documentation. :)

jarrpa commented on roles/openshift_sanitize_inventory/tasks/unsupported.yml:
```diff
@@ -32,3 +32,27 @@
+  fail:
+    msg: |-
+      Configuring a value for openshift_hosted_registry_storage_glusterfs_ips and with a glusterfs_registry host group is not allowed.
+      Specifying a glusterfs_registry host group indicates that a new GlusterFS cluster should be configures, whereas
```
s/configures/configured/

mjudeikis commented:
@jarrpa check if you happy with hosted readme now.

jarrpa commented:
@mjudeikis No. `path` is not somehow distinct from the other variables, and documenting it in a separate table is misleading. The only variable here which is distinct is `ips`, all others apply equally to native and external GlusterFS clusters. Move all variables to the same table, and use the defaults listed in the GlusterFS role (`path` is not a list).

These variables also do not come from the `openshift_storage_glusterfs` role as they are not defined there (merely consumed). The `Additionally` note should go in the GlusterFS role and link back to the hosted role.

mjudeikis commented:
dont merge yet. Doing some testing, dont like what I see...  Will come back.

mjudeikis commented:
@jarrpa all :green_salad: now

mjudeikis commented:
/retest

jarrpa commented:
@mjudeikis That fix makes no sense. Why is only that one instance of the variable changed to use underscores? They should all be the same, and in keeping with the other registry variables they should all use periods.

mjudeikis commented:
@jarrpa , if I set this `openshift.hosted.registry.storage.glusterfs.ips` I get default vaulue for it which is `[]`. Because variable is being set in:
```
openshift_hosted_registry_storage_glusterfs_ips: "{%- set gluster_ips = [] %}{% if groups.glusterfs_registry is defined %}{% for node in groups.glusterfs_registry %}{%- set _ = gluster_ips.append(hostvars[node].glusterfs_ip | default(hostvars[node].openshift.common.ip)) %}{% endfor %}{{ gluster_ips }}{% elif groups.glusterfs is defined %}{% for node in groups.glusterfs %}{%- set _ = gluster_ips.append(hostvars[node].glusterfs_ip | default(hostvars[node].openshift.common.ip)) %}{% endfor %}{{ gluster_ips }}{% else %}{{ openshift.hosted.registry.storage.glusterfs.ips }}{% endif %}"
```
periods value is still default at that state, as we dont have glusterfs_ips variable in the inventory file for default CNS deployment. I didnt looked too deep how our fact plugin works, but as If I understood right "dotted" value is being populated only on the initial fact setting playbook. If variable is set after, like in our case, it does not propagate to dotted value. 

Did the number of different tests and looks we have to use underscored one here, as we are changing its value locally.

jarrpa commented:
.....oh geez, that's terrible. Not your fault, but still terrible. I don't like it, but I'll call fixing it properly out of scope for this PR.

LGTM!

mjudeikis commented:
You don't know how much trouble this "fact plugin" is causing us daily :D why simple default.yml was not enough :D

Thanks, :) time for next one! :D

jarrpa commented:
/lgtm

jarrpa commented:
/lgtm

jarrpa commented:
/kind rfe

mjudeikis commented:
/retest

mjudeikis commented:
/retest

mjudeikis commented:
/retest

jarrpa commented on roles/openshift_storage_glusterfs/tasks/heketi_deploy_part2.yml:
```diff
@@ -1,6 +1,6 @@
 ---
 - name: Create heketi DB volume
-  command: "{{ glusterfs_heketi_client }} setup-openshift-heketi-storage --listfile /tmp/heketi-storage.json"
+  command: "{{ glusterfs_heketi_client }} setup-openshift-heketi-storage --image {{ glusterfs_heketi_image}}:{{ glusterfs_heketi_version }} --listfile /tmp/heketi-storage.json"
```
I beleive the `--image` tag is no longer supported in heketi, as it was just a workaround until it was able to reuse the image specified in the Pod spec. Remove this.

mjudeikis commented:
@jarrpa done

mjudeikis commented:
/test

mjudeikis commented:
/retest

jarrpa commented:
/lgtm

mjudeikis commented:
/retest

mjudeikis commented:
@jarrpa looks like not related test are broken

mjudeikis commented:
/retest

mjudeikis commented:
/retest

mjudeikis commented:
@sdodson @jarrpa we have BZ and issue open for this from my previous client too. Tests look like flakes.

sdodson commented:
@mjudeikis btw, if we want these in v3.7 they need to be cherrypicked back to those branches. I'll leave that up to you and @jarrpa to decide but the bot can automatically cherry-pick if you tell it to via "/cherrypick $branch"

mjudeikis commented:
@sdodson I need to check code state for 3.7 but it should be fine to cherrypick. I would like to do that for this and https://github.com/openshift/openshift-ansible/pull/6073 if @jarrpa is fine with that.

jarrpa commented:
@sdodson @mjudeikis I'm fine with backporting both as "nicetohave"s. I don't want this delaying the current GlusterFS-related errata.

sdodson commented:
you can go ahead and / cherrypick then add / hold to the resultant PRs so they don't merge

jarrpa commented:
Sure thing!

/cherrypick release-3.7

jarrpa commented:
@sdodson Looks like it didn't heed my commands! is it likely I don't have perms for that?

mjudeikis commented:
/cherrypick release-3.7

mjudeikis commented:
@jarrpa I think CI was down. Both PR's raised and on hold for now.

jarrpa commented:
@mjudeikis Cool, thanks.
